### ISSUE TITLE
RuntimeError: Expected object of scalar type Long but got scalar type Byte for argument #2 'target'

### ISSUE THREAD

**Issue by vpapaioannou** [davidtvs/pytorch-lr-finder#19]
With `optimizer = torch.optim.Adam( model.parameters(), lr = learning_rate, weight_decay = weight_decay)`

`criterion = nn.CrossEntropyLoss( weight = None, ignore_index = ignore_index, reduce = False)`

and then executing

`lr_finder = LRFinder(model, optimizer, criterion, device="cuda")
lr_finder.range_test( dataLoader[ 'train'], end_lr=100, num_iter=100)
lr_finder.plot() # to inspect the loss-learning rate graph
lr_finder.reset() # to reset the model and optimizer to their initial state`

I am getting the error,

RuntimeError: Expected object of scalar type Long but got scalar type Byte for argument #2 'target'

Please find below the whole trace.
So far training my models with the above optimizer and criterion I do not have any problem.

![image](https://user-images.githubusercontent.com/6065702/72471348-7aaab100-37b0-11ea-822f-57872cb37a43.png)

**Comment by NaleRaphael** [comment]
Hi @vpapaioannou ,

Could you please check the `dtype` of a single sample inside the `dataloader`? e.g.
```python
print([v.dtype for v in next(iter(dataloader['train']))])
```

It seems that `dtype` of labels in your `dataloader` is `torch.uint8`, you might need to convert them to `torch.int64` or `torch.long`.

**Comment by vpapaioannou** [comment]
After returning the label image as torch.LongTensor( label) and re - running the code, I am getting,

![image](https://user-images.githubusercontent.com/6065702/72651291-edeb2900-3950-11ea-9f1e-d7b89317227d.png)

I am executing the code in a Google Colab notebook and so far I didn't have problems with numpy arrays conversion to Tensors.

Note that for some reason that I don't know, in the training loop I do similar conversions using e.g.

`label.type( 'torch.LongTensor)`

But this same syntax now gives the error,

`AttributeError: 'numpy.ndarray' object has no attribute 'type'`

**Comment by NaleRaphael** [comment]
In the attached image you given, the error message tells you that CUDA is out of memory. Maybe you could reduce the batch size of `dataloader['train']` and retry to see whether OOM issue still exists.

> ... in the training loop I do similar conversions using e.g. `label.type( 'torch.LongTensor)`

According to the error message, `label` is an `numpy.ndarray` and there is no method called `type` could be used.
If you want to perform type casting on `label` by `label.type()`, you have to make sure that `label` is a `torch.Tensor`. Or you can use `label.astype(np.int64)` while `label` is a `numpy.ndarray` and convert it to `torch.Tensor` later.

**Comment by vpapaioannou** [comment]
I decreased batch size to 4 to ensure no memory problems occur. However,

![image](https://user-images.githubusercontent.com/6065702/72842943-89d5a700-3c67-11ea-88f1-482da61f55b1.png)

**Comment by NaleRaphael** [comment]
There might be some problems in your model, because that error message indicates the shape of loss is not `torch.Size([1])`. (see also [this discussion](https://discuss.pytorch.org/t/loss-backward-raises-error-grad-can-be-implicitly-created-only-for-scalar-outputs/12152))

It would be nice for solving this problem if you can provide more information about your training task. e.g. model architecture, is model running on single/multiple GPU (like the link mentioned above) or not...
With a minimal, producible example would be better.

By the way, you can also use [`torch.utils.checkpoint`](https://pytorch.org/docs/stable/checkpoint.html) to train the model. Since batch size affects model performance, you can use this tool without decreasing batch size too much (but it takes more time on training).

**Comment by vpapaioannou** [comment]
Thanks for your prompt response.

I have been running this code for months now on Google Colab with no problem and with` batch size = 256`. I observe that after installing the lr-finder the code breaks. Errors are reported where there was no problem.

In other words, I would suggest that it is a compatibility issue. I would be happy, if I could help somehow on that.

Thanks,

Vassilis

**Comment by NaleRaphael** [comment]
Yeah, I've to say that it's a weird problem and I would like to help you to solve it.
But I'm sorry about that there is currently no sufficient information for us to figure out whether this issue is resulted by LRFinder or not.

I would appreciate it if you could provide a minimal and reproducible example for us to find out the cause of this problem, and even improve this package!

**Comment by vpapaioannou** [comment]
The code that I am running is basically this one, https://github.com/choosehappy/PytorchDigitalPathology/blob/master/segmentation_epistroma_unet/train_unet.ipynb

If you would like to try this code and see what happens, it would be great.

**Comment by pavlinakoutecka** [comment]
Maybe I am little late with this remark - I also encountered a similar error, something like `RuntimeError: Expected object of scalar type Long but got scalar type Float for argument #2 'target'
`. I downloaded the source code and tried to handle this situation. 

After a little adjustment, the official code started to work. I added line `labels = labels.squeeze(1).long() ` to the function _train_batch() in LRFinder class.  The source code now looks something like this:

```

    ...

    def _train_batch(
        self, train_iter, accumulation_steps, non_blocking_transfer=True
    ):
        self.model.train()
        total_loss = None  # for late initialization

        self.optimizer.zero_grad()
        for i in range(accumulation_steps):
            inputs, labels = next(train_iter)
            labels = labels.squeeze(1).long()
            inputs, labels = self._move_to_device(
                inputs, labels, non_blocking=non_blocking_transfer
            )

    ....

```

Hopefully, this will work also for you.

**Comment by davidtvs** [comment]
This kind of issue is usually a sign that the split between the training loop and the data handling on the user side is not well defined. It's common to see training loops doing some sort of data processing/handling/conversions and that should not be the case. Those operations should all be located in the Dataset class so that the input to the model and the labels that come out of the DataLoader can be directly passed to the model and criterion, respectively.

An example of this can be found in the notebook that @vpapaioannou posted. The following lines are found in the training loop:
```py
y_weight = y_weight.type('torch.FloatTensor').to(device)
y = y.type('torch.LongTensor').to(device)
```
The data conversions to `FloatTensor` should be done inside the `Dataset` class and not in the training loop.

In your case, @pavlinakoutecka, I don't have enough information to pinpoint the issue, but I would make an educated guess that your `__getitem__` method in the dataset class returns something like this:
```py
return img, labels
``` 
and the error will go away if you instead return:
```py
return img, labels.squeeze(1).long()
```

In other words, this kind of issue can usually be fixed by moving any data processing/handling/conversions from the training loop to the `__getitem__` method in the dataset class.

**Comment by davidtvs** [comment]
A new release (v0.2.0) is available that addresses issues like this one by allowing the user to wrap the training and validation DataLoader in a `TrainDataLoaderIter` and `TrainDataLoaderIter' and do further data processing/handling/conversions there. For a usage example see: [examples/lrfinder_cifar10_dataloader_iter](https://github.com/davidtvs/pytorch-lr-finder/blob/master/examples/lrfinder_cifar10_dataloader_iter.ipynb).

However, I still think that the best way to deal with this issue is by placing all needed data processing/handling/conversions in the DataLoader as described in my previous comment.

Closing this in the meantime feel free to reopen if not resolved.
